crates/compression/src/compress.rs

```diff
@@ -0,0 +1,242 @@
+                    db: self.db,
+                })
+            }
+        }
+
+        impl<D> CompressCtx<D> where D: CompressDb {
+            /// Finalizes the compression context, returning the changes to the registry.
+            /// Commits the registrations and cache evictor states to the database.
+            fn finalize(mut self) -> anyhow::Result<RegistrationsPerTable> {
+                let mut registrations = RegistrationsPerTable::default();
+                $(
+                    self.$ident.cache_evictor.commit(&mut self.db)?;
+                    for (key, value) in self.$ident.changes.into_iter() {
+                        registrations.$ident.push((key, value));
+                    }
+                )*
+                registrations.write_to_registry(&mut self.db, self.timestamp)?;
+                Ok(registrations)
+            }
+        }
+
+        $(
+            impl<D> CompressibleBy<PrepareCtx<D>> for $type
+            where
+                D: TemporalRegistry<$type> + EvictorDb<$type>
+            {
+                async fn compress_with(
+                    &self,
+                    ctx: &mut PrepareCtx<D>,
+                ) -> anyhow::Result<RegistryKey> {
+                    if *self == <$type>::default() {
+                        return Ok(RegistryKey::ZERO);
+                    }
+                    if let Some(found) = ctx.db.registry_index_lookup(self)? {
+                        let key_timestamp = ctx.db.read_timestamp(&found)
+                            .context("Database invariant violated: no timestamp stored but key found")?;
+                        if ctx.config.is_timestamp_accessible(ctx.timestamp, key_timestamp)? {
+                            ctx.accessed_keys.$ident.insert(found);
+                        }
+                    }
+                    Ok(RegistryKey::ZERO)
+                }
+            }
+
+            impl<D> CompressibleBy<CompressCtx<D>> for $type
+            where
+                D: TemporalRegistry<$type> + EvictorDb<$type>
+            {
+                async fn compress_with(
+                    &self,
+                    ctx: &mut CompressCtx<D>,
+                ) -> anyhow::Result<RegistryKey> {
+                    if self == &Default::default() {
+                        return Ok(RegistryKey::DEFAULT_VALUE);
+                    }
+                    if let Some(found) = ctx.db.registry_index_lookup(self)? {
+                        let key_timestamp = ctx.db.read_timestamp(&found)
+                            .context("Database invariant violated: no timestamp stored but key found")?;
+                        if ctx.config.is_timestamp_accessible(ctx.timestamp, key_timestamp)? {
+                            return Ok(found);
+                        }
+                    }
+                    if let Some(found) = ctx.$ident.changes_lookup.get(self) {
+                        return Ok(*found);
+                    }
```

Comment:
The first part is fixed in 0967f631978d39ea2f83a21dd69b2ed3177d13e8. Not sure what you mean about reusing `accessed_keys` here.

Reply:
I meant in the `PrepareCtx`. We also can avoid calling database by checking `accessed_keys` first

Reply:
Like this? 8787971aa53bd341a013080aa4079499d8d5641a

Reply:
Yep. I also thought that we could maybe have a cache for `self`, but since it is the type that we need to store and requires hashing of the predicates and script, it is better to leave it as it is now=)